crates/ruff_db/src/system/walk_directory.rs

```diff
@@ -0,0 +1,323 @@
+    pub fn path(&self) -> &SystemPath {
+        &self.path
+    }
+
+    /// The full path that this entry represents.
+    /// Analogous to [`DirectoryEntry::path`], but moves ownership of the path.
+    pub fn into_path(self) -> SystemPathBuf {
+        self.path
+    }
+
+    /// Return the file type for the file that this entry points to.
+    ///
+    /// This entry doesn't have a file type if it corresponds to stdin.
+    pub fn file_type(&self) -> FileType {
+        self.file_type
+    }
+
+    /// Return the file name of this entry.
+    ///
+    /// If this entry has no file name (e.g., `/`), then the full path is
+    /// returned.
+    pub fn file_name(&self) -> &str {
+        self.path.file_name().unwrap_or(self.path.as_str())
+    }
+
+    /// Returns the depth at which this entry was created relative to the root.
+    pub fn depth(&self) -> usize {
+        self.depth
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum WalkState {
+    /// Continue walking as normal
+    Continue,
+
+    /// If the entry given is a directory, don't decent into it.
+    /// In all other cases, this has no effect.
+    Skip,
+
+    /// Quit the entire iterator as soon as possible.
+    ///
+    /// Note: This is an inherently asynchronous action. It's possible
+    /// for more entries to be yielded even after instructing the iterator to quit.
+    Quit,
+}
+
+pub struct Error {
+    pub(super) depth: Option<usize>,
+    pub(super) kind: ErrorKind,
+}
+
+impl Error {
+    pub fn depth(&self) -> Option<usize> {
+        self.depth
+    }
+
+    pub fn kind(&self) -> &ErrorKind {
+        &self.kind
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match &self.kind {
+            ErrorKind::Loop { ancestor, child } => {
+                write!(
+                    f,
+                    "File system loop found: {child} points to an ancestor {ancestor}",
+                )
+            }
+            ErrorKind::Io {
+                path: Some(path),
+                err,
+            } => {
+                write!(f, "IO error for operation on {}: {}", path, err)
+            }
+            ErrorKind::Io { path: None, err } => err.fmt(f),
+            ErrorKind::NonUtf8Path { path } => {
+                write!(f, "Non-UTF8 path: {}", path.display())
+            }
+        }
+    }
+}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl std::error::Error for Error {}
+
+#[derive(Debug)]
+pub enum ErrorKind {
+    /// An error that occurs when a file loop is detected when traversing
+    /// symbolic links.
+    Loop {
+        ancestor: SystemPathBuf,
+        child: SystemPathBuf,
+    },
+
+    /// An error that occurs when doing I/O
+    Io {
+        path: Option<SystemPathBuf>,
+        err: std::io::Error,
+    },
```

Comment:
To me it depends on how the enum variants are used. Our AST used to use struct enum variants over named structs. This was very cumbersome in downstream tools because it was impossible to pass an entire `StmtFunctionDef`, because it lacked a name. That's why you still see many lint rules accepting all individual fields of their node instead of the entire node. 

Another motivation for named structs is if you want to lock-down access to individual fields because the enum fields always have the same visibility as the enum itself. 


On the other hand, using variant structs allows Rust to be more clever when it comes to layout optimizations and it's a lot less boilerplate to define the enum and to match on it (no nesting). 


I'm using struct variants here because

a) I think it's unlikely that some code would want to pass the entire variant (have it a named type) and even if, that code would at most have to accept two arguments
b) The main use case is to match on the variant and extract the fields, which is more brief with struct variants.